### ISSUE TITLE
Use kl_divergence() in SVI

### ISSUE THREAD

**Issue by fritzo** [uber/pyro#688]
PyTorch distributions implements a `kl_divergence()` function. We should use this when available in SVI. First we need to account for Pyro distribution wrappers:
```py
def kl_divergence(p, q):
    try:
        p = p.torch_dist
        q = q.torch_dist
    except AttributeError:
        raise NotImplementedError
    return torch.distributions.kl_divergence(p, q)
```
then in SVI we can `try...except`:
```py
assert self.assume_mean_field and self.use_analytic_kl
try:
    kl = kl_divergence(guide_site['fn'], model_site['fn'])
    elbo += kl.detach().sum()
    surrogate_elbo += kl
except NotImplementedError:
    # fall back to monte carlo algorithm
```

**Comment by jacobrgardner** [comment]
Is there any possibility of this working now that some of the KL divergence implementation issues are resolved upstream? 

@gpleiss and I are working on using Pyro's SVI implementation as a possible backend for SVI in GPyTorch, and we are finding that Monte Carlo estimation of the KL divergence is a moderately large performance hit compared to using our custom KL divergence implementation between `MultivariateNormal` instances.

**Comment by fritzo** [comment]
@jacobrgardner Sure we can look into this again. Would you be willing to provide a test case with your desired model structure? @martinjankowiak has concerns that analytic KLs will apply to only a small class of models, and it would help us to see practical examples.

**Comment by jacobrgardner** [comment]
Sure thing! 

Let me try to finish up this variational refactor on our end within the next week or so. That will include support for using Pyro as an alternative SVI implementation and include at least one example notebook of doing the integration (also probably solving #1265). Then we'll have some concrete examples of the both the integration and the performance gap we are seeing that you can look at.

**Comment by martinjankowiak** [comment]
well, it's not so much that i'm concerned that analytic KLs would apply to a small class of models, it's more so that the model/guide have to have a certain dependency structure to allow for an analytic KL to be operable. also it would be a bit tricky to implement in the more general case (e.g. play nice with rao-blackwellization of score function gradient estimators).

a (rather gross) workaround i use in some cases is to create a custom distribution with a custom `log_prob` that includes the KL divergence (effectively folding a particular choice of the guide into the model).

**Comment by eb8680** [comment]
@martinjankowiak now that #1521 is merged, do we have access to enough information in principle to implement the general case of this?

**Comment by martinjankowiak** [comment]
@eb8680 well i don't think we ever lacked the information (at least inside of `TraceGraph`) it's more of a question of making it play nice with everything else (score function gradients, baselines, enumeration, or whatever else there is)

**Comment by royf** [comment]
Very cool, thanks!

Are there plans to support non-reparametrizable variables?
I have a use case where variables cannot be reparametrized, and where training without analytic KL cannot work.
So I cannot use Pyro until it supports analytic KL of non-reparametrizable variables.
